File: 21551-naveen-siva-kumar/week2/1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    //public int left=0,right;
    //public TreeNode r;
    int j=0;
    int size =0;
public : TreeNode* bst(vector<int>& preorder,int left,int right){
    
        //cout<<"left :"<<left<<"right :"<<right<<endl;
        //cout<<"preorder[left] "<<preorder[left]<<"preorder[right] "<<preorder[right]<<endl;
        TreeNode* root = new TreeNode(preorder[left],nullptr,nullptr);   
        //TreeNode* r = root;
        if(left==right) return root;
        int temp = left,mid = -1;
        for(int i=left+1;i<=right;i++){
            
            if(i==right) break;
            
            if((preorder[i]<preorder[temp]) && (preorder[temp]<preorder[i+1])){
                
                mid=i;
                
                j++;
                //cout<<"i :"<<i<<" i+1 :"<<i+1<<"temp"<<temp<<endl;
                //cout<<"left"<<left<<"mid"<<mid<<"right"<<right<<endl;
                //cout<<"mid"<<mid<<endl;
                //cout<<preorder[i]<<" "<<preorder[temp]<<" "<<preorder[i+1]<<endl;
                //cout<<"j "<<j<<endl;
                break;
            }
            
        }
        //cout<<mid<<endl;
        if(mid==-1 && preorder[temp]>preorder[temp+1]){
                //mid=left;
                //cout<<"i :"<<i<<" i+1 :"<<i+1<<"temp"<<temp<<endl;
                //cout<<"left"<<left<<"mid"<<mid<<"right"<<right<<endl;
                //cout<<"mid"<<mid<<endl;
                //cout<<preorder[i]<<" "<<preorder[temp]<<" "<<preorder[i+1]<<endl;
                root->left=bst(preorder,left+1,right);
                //cout<<"left value"<<(root->left)->val;//<<endl;
                //cout<<" for "<<root->val<<endl;
                return root;
            }
            if(mid==-1 && preorder[temp]<preorder[temp+1]){
                //mid=left;
                root->right=bst(preorder,left+1,right);
                return root;
            }
        //cout<<endl;
        //cout<<"left "<<preorder[left+1]<<"mid "<<preorder[mid]<<endl;
        //if(left+1 < this->size){
            if(left+1>mid){
                root->left = nullptr;
            }
            else{ 
                //if(left+1)
                root->left = bst(preorder,left+1,mid);
            }
            // if(root->left!=nullptr){
            //     cout<<"left value"<<(root->left)->val;
            //     cout<<" for "<<root->val<<endl;
            // }
            // else{
            //     cout<<"left value"<<"null";//<<endl;
            //     cout<<" for "<<root->val<<endl;
            // }
            //cout<<endl;
            //cout<<"right "<<preorder[right]<<"mid "<<preorder[mid+1]<<endl;
            if(mid+1>right){
                root->right=nullptr;
            }
            else {
                if(mid+1>this->size){
                    mid=right-1;
                }
                else
                    root->right = bst(preorder,mid+1,right);
            }
            // if(root->right!=nullptr) {      
            //     cout<<"right value"<<(root->right)->val;
            //     cout<<" for "<<root->val<<endl;
            // }
            //     //cout<<endl;
            // else{
            //     cout<<"right value"<<"null";
            //     cout<<" for "<<root->val<<endl;
            // }
        //}
        return root;
        
        
    }
    
    
    
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        this->size=preorder.size();
        return bst(preorder,0,preorder.size()-1);

    }
};

File: 21557 - Mukesh J/Week 2/1008. Construct Binary Search Tree from Preorder Traversal.java
Given an array of integers preorder, which represents the preorder traversal of a BST (i.e., binary search tree), construct the tree and return its root.

It is guaranteed that there is always possible to find a binary search tree with the given requirements for the given test cases.
A binary search tree is a binary tree where for every node, any descendant of Node.left has a value strictly less than Node.val, and any descendant of Node.right has a value strictly greater than Node.val.
A preorder traversal of a binary tree displays the value of the node first, then traverses Node.left, then traverses Node.right.
 

  Example 1:

Input: preorder = [8,5,1,7,10,12]
Output: [8,5,10,1,7,null,12].

  Example 2:

Input: preorder = [1,3]
Output: [1,null,3].
 

  Constraints:

1 <= preorder.length <= 100
1 <= preorder[i] <= 1000
All the values of preorder are unique.


  Data Structure:

Tree, Binary Tree.
  
  Idea:
1. Construct an insert function with root and value to be inserted to the binary search tree.
2. In the insert function, if the root's value is greater than the given value, then traverse to the right side of the tree recursively and insert. 
3. If the root’s value is less than the given value, then traverse to the left side of the tree recursively and insert.
4. If root is null, then return node with value.
5. In the bstFromPreorder function, Each value in the preorder array is sent to BST’s root node for inserting.
6. Once all preorder array elements are inserted, then return root.
7. Time: O(N) where N is the length of the Preorder Array.


  Solution:

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode insert(TreeNode root, int val){
        if(root==null)
            return new TreeNode(val);
        if(val>root.val)
            root.right=insert(root.right,val);
        if(val<root.val) 
            root.left=insert(root.left,val);
        return root;
    }
    
    public TreeNode bstFromPreorder(int[] preorder) {
        TreeNode root = null;
        for(int i=0;i<preorder.length;i++){
            root=insert(root,preorder[i]);
        }
        return root;
    }
}
